File: contrail/contrail-core/src/test/java/org/contrail/actor/core/CoordinatorTest.java
/*
 * Copyright (C)2012 D. Plaindoux.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation; either version 2, or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 */

package org.contrail.actor.core;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import junit.framework.TestCase;

import org.junit.Test;
import org.contrail.actor.event.Request;
import org.contrail.actor.event.Response;
import org.contrail.common.concurrent.Promise;

/**
 * <code>CoordinatorTest</code>
 * 
 * @author Didier Plaindoux
 * @version 1.0
 */
public class CoordinatorTest {

	public class PromiseResponse extends Promise<Object,ActorException> implements Response {
		// Nothing to be done
	}

	public static class A {
		private int value;

		public A(int value) {
			super();
			this.value = value;
		}

		public int getValue() {
			return value;
		}

		public void getError() throws Exception {
			throw new Exception();
		}
	}

	@Test
	public void shouldHaveResponseWithLocalActorInvoked() throws Exception {
		final Coordinator coordinator = new Coordinator();

		final A model = new A(42);
		coordinator.actor("A").bindToObject(model);

		final PromiseResponse response = new PromiseResponse();
		final Request request = new Request("getValue");

		coordinator.invoke("A", request, response);

		TestCase.assertEquals(42, response.getFuture().get(1, TimeUnit.SECONDS));
	}

	@Test(expected = ExecutionException.class)
	public void shouldHaveExceptionWithLocalActorInvoked() throws Exception {
		final Coordinator coordinator = new Coordinator();

		final A model = new A(42);
		coordinator.actor("A").bindToObject(model);

		final PromiseResponse response = new PromiseResponse();
		final Request request = new Request("getError");

		coordinator.invoke("A", request, response);

		TestCase.assertEquals(42, response.getFuture().get(1, TimeUnit.SECONDS));
	}

	@Test
	public void shouldHaveResponseWithSentMessageToLocalActor() throws Exception {
		final Coordinator coordinator = new Coordinator();
		coordinator.start();

		final A model = new A(42);
		coordinator.actor("A").bindToObject(model);

		final PromiseResponse response = new PromiseResponse();
		final Request request = new Request("getValue");

		coordinator.send("A", request, response);

		TestCase.assertEquals(42, response.getFuture().get(1, TimeUnit.SECONDS));
	}

	@Test(expected = ExecutionException.class)
	public void shouldHaveExceptionWithSentMessageToLocalActor() throws Exception {
		final Coordinator coordinator = new Coordinator();
		coordinator.start();

		final A model = new A(42);
		coordinator.actor("A").bindToObject(model);

		final PromiseResponse response = new PromiseResponse();
		final Request request = new Request("getError");

		coordinator.send("A", request, response);

		TestCase.assertEquals(42, response.getFuture().get(1, TimeUnit.SECONDS));
	}

	@Test(expected = TimeoutException.class)
	public void shouldHaveTimeOutWithSentMessageToInactiveLocalActor() throws Exception {
		final Coordinator coordinator = new Coordinator();

		final A model = new A(42);
		coordinator.actor("A").bindToObject(model);

		final PromiseResponse response = new PromiseResponse();
		final Request request = new Request("getValue");

		coordinator.send("A", request, response);

		TestCase.assertNotSame(42, response.getFuture().get(1, TimeUnit.SECONDS));
	}

	@Test
	public void shouldHaveTimeOutAndAResultWithSentMessageToInactiveLocalActor() throws Exception {
		final Coordinator coordinator = new Coordinator();

		final A model = new A(42);
		coordinator.actor("A").bindToObject(model);

		final PromiseResponse response = new PromiseResponse();
		final Request request = new Request("getValue");

		coordinator.send("A", request, response);

		try {
			TestCase.assertNotSame(42, response.getFuture().get(1, TimeUnit.SECONDS));
		} catch (TimeoutException e) {
			// continue
		}

		coordinator.start();

		TestCase.assertEquals(42, response.getFuture().get(1, TimeUnit.SECONDS));
	}


	@Test
	public void shouldHaveAResultWithSentMessageToLateBoundLocalActor() throws Exception {
		final Coordinator coordinator = new Coordinator();
		coordinator.start();

		final A model = new A(42);
		final Actor actor = coordinator.actor("A");

		final PromiseResponse response = new PromiseResponse();
		final Request request = new Request("getValue");

		coordinator.send("A", request, response);

		try {
			TestCase.assertNotSame(42, response.getFuture().get(1, TimeUnit.SECONDS));
		} catch (TimeoutException e) {
			// continue
		}
		
		actor.bindToObject(model);

		TestCase.assertEquals(42, response.getFuture().get(1, TimeUnit.SECONDS));
	}
}

File: contrail-web-client/src/test/js/Actor/core/CoordinatorTest.js
/*
 * Copyright (C)2012 D. Plaindoux.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation; either version 2, or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 */

/*global require, setTimeout*/

require([ "Core/object/jObj", "Core/test/jCC", "Core/client/jLoader", "Actor/jActor", "../common/StoredResponse" ],
    function (jObj, jCC, jLoader, jActor, storedResponse) {
        "use strict";

        // ---------------------------------------------------------

        function A() {
            jObj.bless(this);
            this.a = "a";
        }

        A.prototype.n = function () {
            return "A.n()";
        };

        A.prototype.m = function () {
            throw "A.m()";
        };

        // ---------------------------------------------------------

        jCC.scenario("Check actor indirect successful invocation simulated", function () {
            var coordinator, actor, response;

            jCC.
                Given(function () {
                    coordinator = jActor.coordinator();
                }).
                And(function () {
                    actor = coordinator.actor("A").bindToObject(new A());
                }).
                And(function () {
                    response = storedResponse();
                }).
                When(function () {
                    actor.ask(jActor.event.request("n", []), response);
                }).
                Then(function () {
                    jCC.equal(actor.pendingJobs.length, 1, "A new job has been created");
                }).
                When(function () {
                    coordinator.actorRunner();
                }).
                Then(function () {
                    jCC.equal(coordinator.pendingJobs.length, 1, "Job has been push in executable state");
                }).
                And(function () {
                    jCC.equal(actor.pendingJobs.length, 0, "no new job");
                }).
                When(function () {
                    coordinator.jobRunner();
                }).
                Then(function () {
                    jCC.equal(coordinator.pendingJobs.length, 0, "Job has been executed");
                }).
                And(function () {
                    jCC.equal(response.value(), "A.n()", "Checking response type");
                });
        });

        jCC.scenario("Check actor indirect failed invocation simulated", function () {
            var coordinator, actor, response;

            jCC.
                Given(function () {
                    coordinator = jActor.coordinator();
                }).
                And(function () {
                    actor = coordinator.actor("A").bindToObject(new A());
                }).
                And(function () {
                    response = storedResponse();
                }).
                When(function () {
                    actor.ask(jActor.event.request("m", []), response);
                }).
                Then(function () {
                    jCC.equal(actor.pendingJobs.length, 1, "A new job has been created");
                }).
                When(function () {
                    coordinator.actorRunner();
                }).
                Then(function () {
                    jCC.equal(coordinator.pendingJobs.length, 1, "Job has been push in executable state");
                }).
                And(function () {
                    jCC.equal(actor.pendingJobs.length, 0, "no new job");
                }).
                When(function () {
                    coordinator.jobRunner();
                }).
                Then(function () {
                    jCC.equal(coordinator.pendingJobs.length, 0, "Job has been executed");
                }).
                When(function () {
                    response.value();
                }).
                ThenError(function (e) {
                    jCC.equal(e, "A.m()", "Job has been executed and an exception has been raised");
                    coordinator.stop();
                });
        });

        jCC.scenario("Check actor indirect successful message sent", function () {
            var coordinator, actor, response;

            jCC.
                Given(function () {
                    coordinator = jActor.coordinator();
                    coordinator.start();
                }).
                And(function () {
                    actor = coordinator.actor("A").bindToObject(new A());
                }).
                And(function () {
                    response = storedResponse();
                }).
                When(function () {
                    actor.ask(jActor.event.request("n", []), response);
                }).
                ThenAfter(1000, function () {
                    jCC.equal(response.value(), "A.n()", "Checking response type");
                    coordinator.stop();
                });
        });

        jCC.scenario("Check actor indirect using manager successful message sent", function () {
            var coordinator, response;

            jCC.
                Given(function () {
                    coordinator = jActor.coordinator();
                    coordinator.start();
                }).
                And(function () {
                    coordinator.actor("A").bindToObject(new A());
                }).
                And(function () {
                    response = storedResponse();
                }).
                When(function () {
                    coordinator.ask("A", jActor.event.request("n", []), response);
                }).
                ThenAfter(500, function () {
                    jCC.equal(response.value(), "A.n()", "Checking response type");
                    coordinator.stop();
                });
        });

        jCC.scenario("Check actor indirect using manager successful message sent to first un-started coordinator", function () {
            var coordinator, response;

            jCC.
                Given(function () {
                    coordinator = jActor.coordinator();
                }).
                And(function () {
                    coordinator.actor("A").bindToObject(new A());
                }).
                And(function () {
                    response = storedResponse();
                }).
                When(function () {
                    coordinator.ask("A", jActor.event.request("n", []), response);
                }).
                And(function () {
                    coordinator.start();
                }).
                ThenAfter(500, function () {
                    jCC.equal(response.value(), "A.n()", "Checking response type");
                    coordinator.stop();
                });
        });

        jCC.scenario("Check actor indirect using manager successful message sent to first unbound coordinator", function () {
            var coordinator, response, actor;

            jCC.
                Given(function () {
                    coordinator = jActor.coordinator();
                    coordinator.start();
                }).
                And(function () {
                    actor = coordinator.actor("A");
                }).
                And(function () {
                    response = storedResponse();
                }).
                When(function () {
                    coordinator.ask("A", jActor.event.request("n", []), response);
                }).
                And(function () {
                    actor.bindToObject(new A());
                }).
                ThenAfter(500, function () {
                    jCC.equal(response.value(), "A.n()", "Checking response type");
                    coordinator.stop();
                });
        });

        jCC.scenario("Check actor indirect failed message sent", function () {
            var coordinator, actor, response;

            jCC.
                Given(function () {
                    coordinator = jActor.coordinator();
                    coordinator.start();
                }).
                And(function () {
                    actor = coordinator.actor("A").bindToObject(new A());
                }).
                And(function () {
                    response = storedResponse();
                }).
                When(function () {
                    actor.ask(jActor.event.request("m", []), response);
                }).
                ThenAfter(500, function () {
                    try {
                        response.value();
                        jCC.equal(true, false, "expecting an exception");
                    } catch (e) {
                        jCC.equal(e, "A.m()", "Job has been executed and an exception has been raised");
                    }
                    coordinator.stop();
                });
        });

        jCC.scenario("Check actor indirect using manager failed message sent", function () {
            var manager, response;

            jCC.
                Given(function () {
                    manager = jActor.coordinator();
                    manager.start();
                }).
                And(function () {
                    manager.actor("A").bindToObject(new A());
                }).
                And(function () {
                    response = storedResponse();
                }).
                When(function () {
                    manager.ask("A", jActor.event.request("m", []), response);
                }).
                ThenAfter(500, function () {
                    try {
                        response.value();
                        jCC.equal(true, false, "expecting an exception");
                    } catch (e) {
                        jCC.equal(e, "A.m()", "Job has been executed and an exception has been raised");
                    }
                    manager.stop();
                });
        });

        jCC.scenario("Check undefined actor indirect failed message sent", function () {
            var coordinator, response;

            jCC.
                Given(function () {
                    coordinator = jActor.coordinator();
                    coordinator.start();
                }).
                And(function () {
                    response = storedResponse();
                }).
                When(function () {
                    coordinator.ask("A", jActor.event.request("n", []), response);
                }).
                ThenAfter(500, function () {
                    try {
                        response.value();
                        jCC.equal(true, false, "expecting an exception");
                    } catch (e) {
                        jCC.equal(e.message, "L.actor.not.found", "Actor does not exist");
                    }
                    coordinator.stop();
                });
        });

        jCC.scenario("Check sourced module actor indirect using manager", function () {
            var coordinator, response;

            jCC.
                Given(function () {
                    coordinator = jActor.coordinator();
                    coordinator.start();
                }).
                And(function () {
                    coordinator.actor("AtomicString").bindToSource("./AtomicString.js").onLoad(function () {
                        return require("Atomic.String")("Hello!");
                    });
                }).
                And(function () {
                    response = storedResponse();
                }).
                When(function () {
                    coordinator.ask("AtomicString", jActor.event.request("getValue", []), response);
                }).
                ThenAfter(500, function () {
                    jCC.equal(response.value(), "Hello!", "Actor must respond 'Hello!'");
                    coordinator.stop();
                });
        });

        jCC.scenario("Check sourced simple actor indirect using manager", function () {
            var coordinator, response;

            jCC.
                Given(function () {
                    coordinator = jActor.coordinator();
                    coordinator.start();
                }).
                And(function () {
                    coordinator.actor("SimpleString").bindToSource("./SimpleString.js").onLoad(function () {
                        return SimpleObject;
                    });
                }).
                And(function () {
                    response = storedResponse();
                }).
                When(function () {
                    coordinator.ask("SimpleString", jActor.event.request("getName", []), response);
                }).
                ThenAfter(500, function () {
                    jCC.equal(response.value(), "A simple name", "Actor must respond 'A simple name'");
                    coordinator.stop();
                });
        });

        jCC.scenario("Check sourced module actor indirect using manager failed message sent [New formulation]", function () {
            var coordinator, response;

            jCC.
                Given(function () {
                    coordinator = jActor.coordinator();
                    coordinator.start();
                }).
                And(function () {
                    response = storedResponse();
                }).
                When(function () {
                    jLoader().
                        source("./AtomicString.js").
                        onLoad(function () {
                            coordinator.actor("AtomicString").bindToModule("Atomic.String", [ "Hello!" ]);
                            coordinator.ask("AtomicString", jActor.event.request("getValue", []), response);
                        });
                }).
                ThenAfter(500, function () {
                    jCC.equal(response.value(), "Hello!", "Actor must respond 'Hello!'");
                    coordinator.stop();
                });
        });
    });

